From 5a4a35389becdd9b0c17516888273f0ef41a5040 Mon Sep 17 00:00:00 2001
From: Arkadiusz Hiler <ahiler@codeweavers.com>
Date: Fri, 24 Jun 2022 18:05:33 +0300
Subject: [PATCH] win32u: Don't report cloned monitors in
 EnumDisplayMonitors().

Based on:
ca39b1c22dfa ("user32: Don't report mirrored slave monitors in EnumDisplayMonitors.")
2affb854e524 ("user32: Change slave to a more neutral word.")

Which seem to got lost during:
318673405c62 ("win32u: Move NtUserEnumDisplayMonitors implementation from user32.")

Fixes regression with Elite Dangerous launcher freezing when cloned
displays are present.

Signed-off-by: Arkadiusz Hiler <ahiler@codeweavers.com>
(cherry picked from commit af8ed02b572081206be6c505261f5f2e98a8053c)

CW-Bug-Id: #20759
---
 dlls/win32u/sysparams.c | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/dlls/win32u/sysparams.c b/dlls/win32u/sysparams.c
index 3424de5144c..7e6c2c0bf1e 100644
--- a/dlls/win32u/sysparams.c
+++ b/dlls/win32u/sysparams.c
@@ -220,6 +220,7 @@ struct monitor
     unsigned int flags;
     RECT rc_monitor;
     RECT rc_work;
+    BOOL is_clone;
 };
 
 static struct list adapters = LIST_INIT(adapters);
@@ -1220,7 +1221,7 @@ static BOOL update_display_cache_from_registry(void)
     DWORD adapter_id, monitor_id, monitor_count = 0, size;
     KEY_FULL_INFORMATION key;
     struct adapter *adapter;
-    struct monitor *monitor;
+    struct monitor *monitor, *monitor2;
     HANDLE mutex = NULL;
     NTSTATUS status;
     BOOL ret;
@@ -1265,6 +1266,15 @@ static BOOL update_display_cache_from_registry(void)
                 break;
             }
 
+            LIST_FOR_EACH_ENTRY(monitor2, &monitors, struct monitor, entry)
+            {
+                if (EqualRect(&monitor2->rc_monitor, &monitor->rc_monitor))
+                {
+                    monitor->is_clone = TRUE;
+                    break;
+                }
+            }
+
             monitor->handle = UlongToHandle( ++monitor_count );
             list_add_tail( &monitors, &monitor->entry );
         }
@@ -1871,6 +1881,7 @@ BOOL WINAPI NtUserEnumDisplayMonitors( HDC hdc, RECT *rect, MONITORENUMPROC proc
                                 get_thread_dpi() );
         offset_rect( &monrect, -origin.x, -origin.y );
         if (!intersect_rect( &monrect, &monrect, &limit )) continue;
+        if (monitor->is_clone) continue;
 
         enum_info[count].handle = monitor->handle;
         enum_info[count].rect = monrect;
